services/brig/src/Brig/Team/API.hs

```diff
@@ -230,69 +251,163 @@ createInvitation' tid mUid inviteeRole mbInviterUid fromEmail body = do
-  pure $! Public.InvitationList (DB.resultList rs) (DB.resultHasMore rs)
+  let toInvitations is = mapM (toInvitation showInvitationUrl) is
+  lift (liftSem $ Store.lookupInvitationsPaginated mSize tid startingId) >>= \case
+    PaginatedResultHasMore storedInvs -> do
+      invs <- toInvitations storedInvs
+      pure $ InvitationList invs True
+    PaginatedResult storedInvs -> do
+      invs <- toInvitations storedInvs
+      pure $ InvitationList invs False
+
+-- \| brig used to not store the role, so for migration we allow this to be empty and fill in the
+-- default here.
+toInvitation ::
+  ( Member TinyLog r
+  ) =>
+  ShowOrHideInvitationUrl ->
+  StoredInvitation ->
+  (Handler r) Invitation
+toInvitation showUrl storedInv = do
+  url <- mkInviteUrl showUrl storedInv.teamId storedInv.code
+  pure $
+    Invitation
+      { team = storedInv.teamId,
+        role = fromMaybe defaultRole storedInv.role,
+        invitationId = storedInv.invitationId,
+        createdAt = storedInv.createdAt,
+        createdBy = storedInv.createdBy,
+        inviteeEmail = storedInv.email,
+        inviteeName = storedInv.name,
+        inviteeUrl = url
+      }
+
+mkInviteUrl ::
+  (Member TinyLog r) =>
+  ShowOrHideInvitationUrl ->
+  TeamId ->
+  InvitationCode ->
+  (Handler r) (Maybe (URIRef Absolute))
+mkInviteUrl HideInvitationUrl _ _ = pure Nothing
+mkInviteUrl ShowInvitationUrl team (InvitationCode c) = do
+  template <- invitationEmailUrl . invitationEmail . snd <$> teamTemplates Nothing
+  branding <- view App.templateBranding
+  let url = Text.toStrict $ renderTextWithBranding template replace branding
+  parseHttpsUrl url
+  where
+    replace "team" = idToText team
+    replace "code" = toText c
+    replace x = x
+    parseHttpsUrl :: (Member TinyLog r) => Text -> (Handler r) (Maybe (URIRef Absolute))
+    parseHttpsUrl url =
+      either (\e -> lift . liftSem $ logError url e >> pure Nothing) (pure . Just) $
+        parseURI laxURIParserOptions (encodeUtf8 url)
+    logError url e =
+      Log.err $
+        Log.msg @Text "Unable to create invitation url. Please check configuration."
+          . Log.field "url" url
+          . Log.field "error" (show e)
 
-getInvitation :: (Member GalleyAPIAccess r) => UserId -> TeamId -> InvitationId -> (Handler r) (Maybe Public.Invitation)
+getInvitation ::
+  ( Member GalleyAPIAccess r,
+    Member InvitationCodeStore r,
+    Member TinyLog r
+  ) =>
+  UserId ->
+  TeamId ->
+  InvitationId ->
+  (Handler r) (Maybe Public.Invitation)
 getInvitation uid tid iid = do
-  ensurePermissions uid tid [AddTeamMember]
-  showInvitationUrl <- lift $ liftSem $ GalleyAPIAccess.getExposeInvitationURLsToTeamAdmin tid
-  lift $ wrapClient $ DB.lookupInvitation showInvitationUrl tid iid
+  invitationM <- lift . liftSem $ Store.lookupInvitation tid iid
+  case invitationM of
+    Nothing -> pure Nothing
+    Just invitation -> do
+      ensurePermissions uid tid [AddTeamMember]
```

Comment:
moving this here changes behavior in case permission is not granted *and* the invitation is not available.

why change it in the first place?  if this is about runtime performance: please don't :)

Reply:
I’m sure it wasn’t runtime performance, probably needs to get looked at if this changes behaviour.

Reply:
Yeah it changes the behaviour, before even if there were no invitations, we'd get an error on permission failing, now we're leaking information.